meilisearch-lib/src/auth_controller/auth_store.rs

```diff
@@ -0,0 +1,257 @@
+        }
+    }
+
+    fn delete_key_from_inverted_db(
+        &self,
+        wtxn: &mut RwTxn,
+        key: &[u8; KEY_ID_LENGTH],
+    ) -> Result<()> {
+        let mut iter = self
+            .action_keyid_index_expiration
+            .remap_types::<ByteSlice, DecodeIgnore>()
+            .prefix_iter_mut(wtxn, key)?;
+        while let Some(_) = iter.next().transpose()? {
+            // safety: we don't keep references from inside the LMDB database.
+            unsafe { iter.del_current()? };
+        }
+
+        Ok(())
+    }
+}
+
+/// Heed codec allowing to encode/decode everithing that implement Serialize and Deserialize
+/// in order to store it in heed.
+/// This is obviously not the best approach and should never be used for big and numerous objects,
+/// but it is a simple one.
+pub struct SerdeJsonCodec<T>(std::marker::PhantomData<T>);
+
+impl<'a, T> heed::BytesDecode<'a> for SerdeJsonCodec<T>
+where
+    T: Deserialize<'a> + 'a,
+{
+    type DItem = T;
+
+    fn bytes_decode(bytes: &'a [u8]) -> Option<Self::DItem> {
+        serde_json::from_slice(bytes).ok()
+    }
+}
+
+impl<'a, T> heed::BytesEncode<'a> for SerdeJsonCodec<T>
+where
+    T: Serialize + 'a,
+{
+    type EItem = T;
+
+    fn bytes_encode(item: &Self::EItem) -> Option<Cow<[u8]>> {
+        serde_json::to_vec(item).map(|bytes| Cow::Owned(bytes)).ok()
+    }
+}
+
+pub struct ActionKeyIdCodec;
+
+impl<'a> heed::BytesDecode<'a> for ActionKeyIdCodec {
+    type DItem = ([u8; KEY_ID_LENGTH], Action, Option<&'a [u8]>);
+
+    fn bytes_decode(bytes: &'a [u8]) -> Option<Self::DItem> {
+        let (key_id, action_bytes) = try_split_array_at(bytes)?;
+        let (action_bytes, index) = match try_split_array_at(action_bytes)? {
+            (action, []) => (action, None),
+            (action, index) => (action, Some(index)),
+        };
+        let action = Action::from_repr(u8::from_be_bytes(action_bytes))?;
+
+        Some((key_id, action, index))
+    }
+}
+
+impl<'a> heed::BytesEncode<'a> for ActionKeyIdCodec {
+    type EItem = (&'a [u8; KEY_ID_LENGTH], &'a Action, Option<&'a [u8]>);
+
+    fn bytes_encode((key_id, action, index): &Self::EItem) -> Option<Cow<[u8]>> {
+        let mut bytes = Vec::new();
+
+        bytes.extend_from_slice(*key_id);
+        let action_bytes = u8::to_be_bytes(action.repr());
+        bytes.extend_from_slice(&action_bytes);
+        if let Some(index) = index {
+            bytes.extend_from_slice(index);
+        }
+
+        Some(Cow::Owned(bytes))
+    }
+}
+
+/// Divides one slice into two at an index, returns `None` if mid is out of bounds.
+pub fn try_split_at<T>(slice: &[T], mid: usize) -> Option<(&[T], &[T])> {
+    if mid <= slice.len() {
+        Some(slice.split_at(mid))
+    } else {
+        None
+    }
+}
+
+/// Divides one slice into an array and the tail at an index,
+/// returns `None` if `N` is out of bounds.
+pub fn try_split_array_at<T, const N: usize>(slice: &[T]) -> Option<([T; N], &[T])>
```

Comment:
any reason why you need to own that slice here?

Reply:
clever trick btw

Reply:
Just making sure that's a constant size, I don't really need to own it.

Reply:
wouldn't that work?
```rust
pub fn try_split_array_at<T, const N: usize>(slice: &[T]) -> Option<(&[T; N], &[T])> {
    let (head, tail) = try_split_at(slice, N)?;
    let head = head.try_into().ok()?;
    Some((head, tail))
}
```